controller/lib/hal/hal_stm32.h

```diff
@@ -0,0 +1,400 @@
+/* Copyright 2020, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+This file implements the HAL (Hardware Abstraction Layer) for the
+STM32L452 processor used on the controller.  Details of the processor's
+peripherals can be found in the reference manual for that processor:
+   https://www.st.com/resource/en/reference_manual/dm00151940-stm32l41xxx42xxx43xxx44xxx45xxx46xxx-advanced-armbased-32bit-mcus-stmicroelectronics.pdf
+
+Details specific to the ARM processor used in this chip can be found in
+the programmer's manual for the processor available here:
+   https://www.st.com/resource/en/programming_manual/dm00046982-stm32-cortexm4-mcus-and-mpus-programming-manual-stmicroelectronics.pdf
+
+*/
+
+#ifndef DEF_INC_STM32
+#define DEF_INC_STM32
+
+#if !defined( BARE_STM32 )
+#error "the header stm32.h is part of the HAL for the BARE_STM32 build.  It should only be used as part of that build"
+#endif
+
+#include <stdint.h>
+
+#define CPU_FREQ_MHZ       80
+#define CPU_FREQ           (CPU_FREQ_MHZ * 1000000)
+
+// Some useful inline functions to enable/disable interrupts
+inline void IntDisable(){
+   asm volatile( "cpsid i" );
+}
+
+inline void IntEnable(){
+   asm volatile( "cpsie i" );
+}
+
+inline void IntRestore( bool p ){
+   if( p ) IntEnable();
+}
+
+inline bool IntSuspend(){
+   int ret;
+   asm volatile( "mrs   %[output], primask\n\t" "cpsid i": [output] "=r" (ret) );
```

Comment:
This is not sufficiently scary to me.  Can we mark this as "reads/writes memory" so that all loads/stores have to be flushed?  Same for disabling interrupts.

Reply:
I think the volatile keyword after asm is supposed to do that, but I'm happy adding more.
Would that be "=m" in the clobber list?  I'm not entirely certain of the syntax here.

Reply:
> I think the volatile keyword after asm

You would think, but that's not how I read the gcc docs.  Last example of https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html#Volatile suggests to me that reads and writes can be reordered with respect to asm volatile.

I think it's "memory" in the clobber list.

Reply:
OK, thanks.  I added "memory" to the clobber list of the relevant functions and double checked the compiler output by stepping through the assembly language and everything looks good.
Incidentally, it turns out that the the visual studio code debugger is very broken when it comes to stepping through assembly code, at least on the STM32 processor.  I had to manually do it through the debug console which was pretty tedious!